File: leetcode-js/207-course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    'use strict';

    const graph = [];
    const degree = [];

    for (let i = 0; i < numCourses; i += 1) {
        degree[i] = 0;
        graph[i] = [];
    }

    for (const prerequisite of prerequisites) {
        const next = prerequisite[0];
        const prer = prerequisite[1];
        degree[next] += 1;
        graph[prer].push(next);
    }

    const isolated = [];

    for (const entry of degree.entries()) {
        const i = entry[0];
        const count = entry[1];
        if (count === 0) {
            isolated.push(i);
        }
    }

    let count = 0;

    while (isolated.length) {
        count += 1;
        const course = isolated.pop();
        for (const next of graph[course]) {
            degree[next] -= 1;
            if (degree[next] === 0) {
                isolated.push(next);
            }
        }
    }

    return count === numCourses;
};

console.log(canFinish(8,
  [
    [1,0],
    [2,1],
    [0,3],
    [3,1]
  ]));

File: leetcode-swift/207-course-schedule.swift
class Solution {
    func canFinish(numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
        var degree = [Int](count: numCourses, repeatedValue: 0)
        var graph = [[Int]](count: numCourses, repeatedValue: [])

        for prerequisite in prerequisites {
            degree[prerequisite[0]] += 1
            graph[prerequisite[1]].append(prerequisite[0]);
        }

        var isolated = [Int]()

        for (i, count) in degree.enumerate() {
            if count == 0 {
                isolated.append(i)
            }
        }

        var count = 0;

        while isolated.count > 0 {
            count += 1
            let course = isolated.removeLast()
            for next in graph[course] {
                degree[next] -= 1
                if degree[next] == 0 {
                    isolated.append(next)
                }
            }
        }

        return count == numCourses;
    }
}
